src/codeflare_sdk/jobs/jobs.py

```diff
@@ -0,0 +1,135 @@
+
+import abc
+from typing import List
+
+from ray.job_submission import JobSubmissionClient
+from torchx.components.dist import ddp
+from torchx.runner import get_runner, Runner
+from torchx.specs import AppHandle, parse_app_handle
+
+from .config import JobConfiguration
+
+import typing
+if typing.TYPE_CHECKING:
+    from ..cluster.cluster import Cluster
+
+all_jobs: List["Job"] = []
+torchx_runner: Runner = get_runner()
+
+class JobDefinition(metaclass=abc.ABCMeta):
+    """
+    A job definition to be submitted to a generic backend cluster.
+    """
+
+    def submit(self, cluster: "Cluster"):
+        """
+        Method for creating a job on a specific cluster
+        """
+        pass
+
+class Job(metaclass=abc.ABCMeta):
+    """
+    An abstract class that defines the necessary methods for authenticating to a remote environment.
+    Specifically, this class `status` and a `logs` function.
+    """
+
+    def status(self):
+        """
+        Method for retrieving the job's current status.
+        """
+        pass
+
+    def logs(self):
+        """
+        Method for retrieving the job's logs.
+        """
+
+
+class TorchXJobDefinition(JobDefinition):
+    def __init__(self, config: JobConfiguration):
+        """
+        Create the TorchXJob object by passing in a JobConfiguration
+        (defined in the config sub-module).
+        """
+        self.config = config
+
+    def submit(self, cluster: "Cluster") -> "TorchXRayJob":
+        """
+        Submit the job definition to a specific cluster, resulting in a Job object.
+        """
+        return TorchXRayJob(self, cluster)
+
+
+class TorchXRayJob(Job):
+    """
+    Active submission of a dist.ddp job to a Ray cluster which can be used to get logs and status.
+    """
+    def __init__(self, job_definition: TorchXJobDefinition, cluster: "Cluster", *script_args):
+        """
+        Creates job which maximizes resource usage on the passed cluster.
+        """
+        self.job_definition: TorchXJobDefinition = job_definition
+        self.cluster: "Cluster" = cluster
+        j = f"{cluster.config.max_worker}x{max(cluster.config.gpu, 1)}"  # # of proc. = # of gpus
+        _app_handle: AppHandle = torchx_runner.run(
+            app=ddp(
+                *script_args,
+                script = job_definition.config.script,
+                m=job_definition.config.m,
+                name=job_definition.config.name,
+                h=None,  # for custom resource types
+                cpu=cluster.config.max_cpus,
+                gpu = cluster.config.gpu,
+                memMB = 1024 * cluster.config.max_memory,  # cluster memory is in GB
+                j=j,
+                env=job_definition.config.env,
+                debug=job_definition.config.debug  # TODO: expose
+                # max_retries=0,  # default
+                # mounts=None,  # default
+            ),
+            scheduler="ray",  # can be determined by type of cluster if more are introduced
```

Comment:
Should we add this as configurable parameter, "scheduler" in the Config, with 'ray' as the default. And then rename this class `TorchXJob()`? I know the status() and logs() functions assume ray at the moment, but the class name and hard coding  feel a bit awkward here. WDYT?

Reply:
Yeah, the scheduler is really dependent on the cluster, so I don't think it needs to be exposed to the user. It shouldn't be hard coded though.